feat(watch): test run cancelling, feat: `--bail` option for cancelling test run

AriPerkkio opened the pull request:
Contains fixes for two separate issues. This PR can be merged with rebase to preserve the commits. 

- Fixes https://github.com/vitest-dev/vitest/issues/3128
- Fixes https://github.com/vitest-dev/vitest/issues/1459


<details>
   <summary>Video: Watch mode cancel</summary>
   
https://user-images.githubusercontent.com/14806298/230854177-9947981a-2aeb-4992-8387-40ea2f08ab36.mov

</details>

<details>
   <summary>Video: --bail option</summary>


https://user-images.githubusercontent.com/14806298/232055152-3293b568-13b8-4ce8-8c4f-d0c751342223.mov

</details>


To test:

- [x] Multithread
- [x] Single thread
- [x] Child process
- [x] Browser

And each combination with
- [x] Isolation on/off
- [x] TTY on/off

sheremet-va commented on packages/runner/src/run.ts:
```diff
@@ -105,6 +105,14 @@ async function callCleanupHooks(cleanups: HookCleanupCallback[]) {
 }
 
 export async function runTest(test: Test, runner: VitestRunner) {
+  // @ts-expect-error untyped global
```
runner cannot depend on private vitest properties, everything should be provided with the runner class

sheremet-va replied:
I don't know the best solution. We can also add a method there specifically for canceling, I suppose.

AriPerkkio replied:
Alright I'm planning a `Runner.onCancel` method that runners can use to decide how they should cancel the runs. This seems to work nice for `worker_threads` and `child_process` pools but I'll need to figure out browser next. 

```ts
class CustomRunner {
  private cancelRun = false;

  /**
   * Called when test runner should cancel next test runs.
   * Runner should listen for this method and mark tests and suites as skipped in
   * "onBeforeRunSuite" and "onBeforeRunTest" when called.
   */
  onCancel() {
    this.cancelRun = true;
  }

  async onBeforeRunTest(test: Test) {
    if (this.cancelRun) test.mode = "skip";
    // ...
  }

  onBeforeRunSuite(suite: Suite) {
    if (this.cancelRun)
      suite.mode = 'skip'
    // ...
  }
}
```

jgoux commented:
I'm really grateful you're integrating `--bail` into this!

Seeing the demo, is there a possibility to add an option to bail from all the files being tested? So if we test files a, b and c and b is in error, it would stop both a, b and c. This is important to save CI time when we launch everything together. 😁

AriPerkkio commented:
Yes, that is the current behaviour. 

When a test case fails, the runner will ask main thread the count of currently failed test cases (across all test files) and decides if a bail/cancel should be done. When it sees that enough test cases have failed (`failedCount >= config.bail`):
- runner won't start any new **test cases** from the current test file
- runner informs main thread that bail/cancel should be done
  - main thread won't start any new **test files** in the pool
  - main thread / pool will inform all currently active runners (threads) that they should not start any new **test cases** in their current files
  
In practice the test cancelling means that no new tests should be started. It won't forcefully terminate any on-going test cases as this could cause unexpected side effects. The demo video demonstrates this by using 3000ms slow test cases which might be a bit overkill when highlighting the feature. But it will be identical with Jest, I think.

ghiscoding commented:
Just to give my point of view in regards to the issue #3128 which I think triggered the opening of this PR (which I'm also highly anticipating). My use case is that I have to use `--no-threads` because I need child process in [Lerna-Lite](https://github.com/lerna-lite/lerna-lite), however with this option it is extremely slow on a Windows platform, it takes 2-3min to complete all tests before I can start using the test pattern of the CLI (by typing "p" in the shell). I didn't have to wait until the end in Jest and that is what I was expecting Vitest to also do (not yet but with this PR it will)... 

long story short, this will be super helpful for long process execution which is especially worst on Windows with `--no-threads`. I also agree with @AriPerkkio that if any tests were started then they should finish their execution but no new tests should start since we interrupted the process which is also what Jest is doing.

out of topic, `--no-threads` is extremely slow on Windows but the equivalent `runInBand` on Jest wasn't that slow. It's so damn slow that I had switch to WSL (Windows Subsystem for Linux) to get back to a decent test time. I did find a few opened issues about this subject but no real answer on to fix this on Windows, any pointer or future PR would be nice.

Cheers and thanks for opening this PR :)

Aslemammad commented on packages/vitest/src/node/pools/threads.ts:
```diff
@@ -123,6 +141,12 @@ export function createThreadsPool(ctx: Vitest, { execArgv, env }: PoolProcessOpt
+      // If not, remove whole signal and rely on runner level cancellation instead.
+      const signal = new EventEmitter()
+      signal.setMaxListeners(ctx.state.getFiles().length)
+      ctx.onCancel(() => signal.emit('abort'))
```
Yes, v0.5.0 published seconds ago.

AriPerkkio replied:
Awesome, thanks! 🙌

AriPerkkio commented:
Ready for review. @sheremet-va as this PR contains two separate features I'll keep the commits ready for rebasing when merging to main - the release changelog will show both features then. 

The `test-browser-safari` CI build failure also happens on main branch. I think it's related to latest `pnpm@8` update. I saw same error locally but fixed it by clearing `packages/*/node_modules`.

sheremet-va reviewed (changes requested):
LGTM, but can we add more documentation for the `--bail` feature (and add to CLI args)?

AriPerkkio commented:
Added flag to CLI argument documentation page.

Also updated the option documentation to describe its use case in more detail. I'm not sure what more to add - writing documentation is hard. 🙃 

https://github.com/vitest-dev/vitest/blob/b046f0c7ac811b36d1c53fe652c5e13637390751/docs/config/index.md?plain=1#L1379-L1387

capaj commented:
the default is not working. I just got this on CI:

```
$ TZ=UTC NODE_ENV=test vitest run --coverage --bail
file:///home/runner/work/secret/graphql/node_modules/cac/dist/index.mjs:413
          throw new CACError(`option \`${option.rawName}\` value is missing`);
                ^

CACError: option `--bail <number>` value is missing
    at Command.checkOptionValue (file:///home/runner/work/secret/graphql/node_modules/cac/dist/index.mjs:413:17)
    at CAC.runMatchedCommand (file:///home/runner/work/secret/graphql/node_modules/cac/dist/index.mjs:599:13)
    at CAC.parse (file:///home/runner/work/secret/graphql/node_modules/cac/dist/index.mjs:537:12)
    at file:///home/runner/work/secret/graphql/node_modules/vitest/dist/cli.js:66:5
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)

Node.js v18.16.0
```

AriPerkkio commented:
Default is the value that is used when you don't specify `bail` option at all.

When you wish to change `bail` option, specify a value - exactly as the error message there is instructing.

capaj commented:
so I need `--bail 1`, ok thanks

gajus commented:
I don't believe `--bail` works as it is supposed to

https://github.com/vitest-dev/vitest/issues/3372
